src/librustc_front/visit.rs

```diff
@@ -203,26 +203,31 @@ pub fn walk_item<'v, V: Visitor<'v>>(visitor: &mut V, item: &'v Item) {
         ItemExternCrate(..) => {}
         ItemUse(ref vp) => {
             match vp.node {
-                ViewPathSimple(ident, ref path) => {
-                    visitor.visit_ident(vp.span, ident);
+                ViewPathSimple(_ident, ref path) => {
                     visitor.visit_path(path, item.id);
                 }
                 ViewPathGlob(ref path) => {
                     visitor.visit_path(path, item.id);
                 }
                 ViewPathList(ref prefix, ref list) => {
-                    for id in list {
-                        match id.node {
-                            PathListIdent { name, .. } => {
-                                visitor.visit_ident(id.span, name);
+                    let mut path = prefix.clone();
+                    for item in list {
+                        match item.node {
+                            PathListIdent { name, id, .. } => {
+                                path.span = item.span;
+                                path.segments.push(PathSegment {
+                                    identifier: name,
+                                    parameters: PathParameters::none()
+                                });
+                                visitor.visit_path(unsafe{&*(&path as *const _)}, id);
```

Comment:
visit_path is defined as `fn visit_path(&mut self, path: &'v Path, _id: NodeId) {`, i.e. the lifetime of the path is tied to the the lifetime of the visitor, not the lifetime of the callback.  Also, the current PR visits both the path's components and the path as a whole.  Also, cloning the prefix is a bunch of extra computation which is unlikely to be useful 

Probably the most straightforward thing to do would be to add a method `visit_path_list_item(&mut self, prefix: &'v Path, item: &'v PathListItem)` to Visitor and let the callee do the right thing.

Related issue: we probably also want a method on Visitor to visit the prefix; it would be useful for fixing the handling of `use std::foobar::{};`.